cluster/assemblestate/transport.go

```diff
@@ -0,0 +1,495 @@
+// cluster assembly communication. The transport tracks message statistics using
+// atomic counters for monitoring purposes.
+type HTTPSTransport struct {
+	stats TransportStats
+}
+
+// NewHTTPSTransport creates a new [HTTPSTransport] instance with default
+// settings. The returned transport can be used to both serve incoming assembly
+// requests and create clients for outbound communication.
+func NewHTTPSTransport() *HTTPSTransport {
+	return &HTTPSTransport{}
+}
+
+// Serve implements the Transport interface. It starts an HTTPS server with the
+// given [net.Listener] using the provided TLS certificate. The server routes
+// incoming assembly protocol messages to the appropriate handlers.
+//
+// The server handles the following endpoints:
+//   - /assemble/auth: Authentication messages from untrusted peers
+//   - /assemble/routes: Route information from trusted peers
+//   - /assemble/unknown: Device queries from trusted peers
+//   - /assemble/devices: Device identity responses from trusted peers
+//
+// The server runs until the context is cancelled.
+func (t *HTTPSTransport) Serve(ctx context.Context, ln net.Listener, cert tls.Certificate, pa PeerAuthenticator) error {
+	mux := http.NewServeMux()
+	mux.Handle("/assemble/auth", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		t.handleAuth(w, r, pa)
+	}))
+	mux.Handle("/assemble/routes", t.trustedHandler(t.handleRoutes, pa))
+	mux.Handle("/assemble/unknown", t.trustedHandler(t.handleUnknown, pa))
+	mux.Handle("/assemble/devices", t.trustedHandler(t.handleDevices, pa))
+
+	server := &http.Server{
+		Handler: mux,
+		BaseContext: func(l net.Listener) context.Context {
+			return ctx
+		},
+		ErrorLog: log.New(io.Discard, "", 0),
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		listener := tls.NewListener(ln, &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientAuth:   tls.RequireAnyClientCert,
+		})
+
+		// serve always returns a non-nil error, nothing to handle here
+		_ = server.Serve(listener)
+	}()
+
+	// wait for context cancellation
+	<-ctx.Done()
+
+	// shutdown the server
+	_ = server.Close()
+	wg.Wait()
+
+	return nil
+}
+
+type countingReader struct {
+	r     io.Reader
+	count int64
+}
+
+func (c *countingReader) Read(p []byte) (int, error) {
+	n, err := c.r.Read(p)
+	c.count += int64(n)
+	return n, err
+}
+
+func (t *HTTPSTransport) trustedHandler(next func(http.ResponseWriter, *http.Request, VerifiedPeer), pa PeerAuthenticator) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if r.TLS == nil || len(r.TLS.PeerCertificates) != 1 {
+			w.WriteHeader(403)
+			return
+		}
+
+		cert := r.TLS.PeerCertificates[0].Raw
+		peer, err := pa.VerifyPeer(cert)
+		if err != nil {
+			logger.Debug("dropping message from untrusted peer")
+			w.WriteHeader(403)
+			return
+		}
+
+		next(w, r, peer)
+	}
+}
+
+func (t *HTTPSTransport) handleAuth(w http.ResponseWriter, r *http.Request, pa PeerAuthenticator) {
```

Comment:
nitpick: I would have expected this to be defined before trustedHandler, looking at the order of definition/use